mava/utils/jax_training_utils.py

```diff
@@ -58,13 +65,23 @@ def compute_running_mean_var_count(
     delta = batch_mean - mean
     tot_count = count + batch_count
 
-    new_mean = mean + delta * batch_count / tot_count
+    mean = mean + delta * batch_count / tot_count
     m_a = var * count
     m_b = batch_var * batch_count
     M2 = m_a + m_b + jnp.square(delta) * count * batch_count / tot_count
-    new_var = M2 / tot_count
+    var = M2 / tot_count
+    std = jnp.sqrt(var)
     new_count = tot_count
-    new_std = jnp.sqrt(new_var)
+
+    # This assumes the all the features we don't want to
+    # normalise are all always at the front.
+    new_mean = jnp.zeros_like(mean)
+    new_var = jnp.zeros_like(var)
+    new_std = jnp.ones_like(std)
+
+    new_mean = new_mean.at[start_axes:].set(mean[start_axes:])
+    new_var = new_var.at[start_axes:].set(var[start_axes:])
+    new_std = new_std.at[start_axes:].set(std[start_axes:])
```

Comment:
Didn't we say we wanted this to be a slice and not a single number? What if the observations we want to normalize are between observations we don't want to normalize?

So instead of start axes I would have something like `normalization_indicies` where the user could pass in something like `slice(2,10)` or just a numpy array of the indices required to be normalised.

Reply:
I checked and the features we don't want to normalise agent ids and agent previous actions are always added at the front of the the features array.  Given that the agent_ids and actions are always added to the front. We need to adopt a convention so that any other features we don't want to normalise are always at the front. Because we may have issues like we have features maybe from slice(10, 20) we don't want to normalise.  Then the user wraps this environment with concatenate agent ids the adds the agent one hot encoded vectors at the front. Let say we have 5 agents and that the initial number of features was 40. So with concatenate agent  ids now we have 45 features. Also we don't want to normalise from Slice (0, 5) for agent ids and now from Slice (15, 25). This will become very trick to maintain.

Reply:
I see what you are saying, but I think solution I proposed adds more options in the future.  
Ideally we will keep putting things we do not want to normalize at the front and as such getting the indicies will be simple, but it is possible at some in the future that we need to not normalize something in the middle of the observations and with the current method it would not be possible.  
With the proposed method you could pass in [5,6,...,14,15,25,26...] and you would be able to skip normalizing the beginning and the middle of the observations. While you are correct in saying that it would be hard to maintain an env that has something we don't want to normalize in the middle I would rather have the option to do this in the future if we need to, while keeping convention that non-normalized observations stay at the front. 

I think that keeps the maintainability and leaves room for customization if needed, what do you think?

Also @DriesSmit and @RuanJohn what are your thoughts?

Reply:
I like @sash-a's idea :slightly_smiling_face: It keeps it more general. If we wrap the environment the values, that should be normalised, should automatically be different in the same way `obs_normalisation_start` would be different.

Reply:
@sash-a @DriesSmit let's use this approach. We should chat about it more at the standup.

Reply:
I like the future-proofing that your idea adds @sash-a. Along with the slices, can we set the default to be a slice that is inferred from the observations specs? That way, by default the slice could be something like `[-obs_dim:]` without a user having to explicitly pass in a slice. The user can just turn observation normalisation on. In the case of more complex use cases, like what we are future-proofing for, a user could then pass in a more complex slice.

Reply:
@DriesSmit @sash-a @RuanJohn I have made changes to use slices. We still keep the start_index mainly for concatenate_agent_id and previous_options but the use can pass a list for other axes he doesn't want to normalise. The implementation for now assumes 1D features spaces, so it will through error if we if the observation shape has more than 1 dimension.